src/chain/sync/grandpa_warp_sync.rs

```diff
@@ -0,0 +1,327 @@
+use crate::{
+    chain::chain_information::{
+        babe_fetch_epoch, BabeEpochInformation, ChainInformation, ChainInformationConsensus,
+        ChainInformationFinality,
+    },
+    executor::{
+        host::{HostVmPrototype, NewErr},
+        vm::ExecHint,
+    },
+    finality::{grandpa::warp_sync, justification::verify},
+    header::Header,
+    libp2p::PeerId,
+    network::service::{ChainNetwork, GrandpaWarpSyncRequestError},
+};
+use core::ops::{Add, Sub};
+use core::time::Duration;
+
+/// Problem encountered during a call to [`grandpa_warp_sync`].
+#[derive(Debug, derive_more::Display)]
+pub enum Error {
+    #[display(fmt = "All requests failed.")]
+    AllRequestsFailed,
+    #[display(fmt = "{}", _0)]
+    Verifier(verify::Error),
+    #[display(fmt = "{}", _0)]
+    BabeFetchEpoch(babe_fetch_epoch::Error),
+    #[display(fmt = "{}", _0)]
+    NewRuntime(NewErr),
+}
+
+/// The configuration for [`grandpa_warp_sync`].
+pub struct Config<'a, TNow, TPeer, TConn> {
+    pub connected_peers: Vec<PeerId>,
+    pub network: &'a ChainNetwork<TNow, TPeer, TConn>,
+    pub now: TNow,
+    pub chain_index: usize,
+    pub begin_hash: [u8; 32],
+    pub genesis_chain_information: &'a ChainInformation,
+}
```

Comment:
I'm not sure what the best way to tidy up the `now`, `chain_index`, and `begin_hash` fields would be.